### PR TITLE
CR-1120043 : Enhanced xbutil2 to show core information which is not enabled but buffer exists

### PR DESCRIPTION
**Pull request by saifuddin-xilinx**
This PR cover both the following CRs:
[CR-1120043] Missing "locks" section in aie_status_edge.json when Core is unused but buffers exists
[CR-1114191] Please remove the sentence "Pl Kernel Instances in Graph:" from xbutil report.

In AIE, there are some cores which is not enabled but memory buffer exists on those cores.
In AIE metadata "dma_columns", "dma_row", which tells about all the cores where memory buffer present.

We traverse all of that and show the lock, DMA information of those cores, even though those cores are not enable.

No output change for xbutil. It will additionally show more Cores.

### CODE REVIEW

File: src/runtime_src/core/common/info_aie.cpp [hunk 1]
```diff
@@ -531,6 +531,52 @@ populate_aie_core_gmio(const boost::property_tree::ptree& pt, boost::property_tr
   pt_array.add_child("gmios",gmio_array);
 }
 
+// This funtion check the duplicate entry in tile_aaray for the same core [row:col]
```
Comment (rozumx) [inline src/runtime_src/core/common/info_aie.cpp:534]:
```suggestion
// This function check the duplicate entry in tile_aaray for the same core [row:col]
```

File: src/runtime_src/core/common/info_aie.cpp [hunk 2]
```diff
@@ -531,6 +531,52 @@ populate_aie_core_gmio(const boost::property_tree::ptree& pt, boost::property_tr
...
+      if ((tile.second.get<int>("column") == col)
+	  && (tile.second.get<int>("row") == row)) {
+        return true;
+    }
+  }
+
+  return false;
+}
+
+// This funtion populate a specific AIE core which is unused but memory buffers exists [row:col]
```
Comment (rozumx) [inline src/runtime_src/core/common/info_aie.cpp:548]:
```suggestion
// This function populate a specific AIE core which is unused but memory buffers exists [row:col]
```

File: src/runtime_src/core/common/info_aie.cpp [hunk 3]
```diff
@@ -531,6 +531,52 @@ populate_aie_core_gmio(const boost::property_tree::ptree& pt, boost::property_tr
...
+void
+populate_buffer_only_cores(const boost::property_tree::ptree& pt,
+			  const boost::property_tree::ptree& core_info, int gr_id,
+			  boost::property_tree::ptree& tile_array)
+{
+  boost::property_tree::ptree empty_pt;
```
Comment (rozumx) [inline src/runtime_src/core/common/info_aie.cpp:554]:
```suggestion
  const boost::property_tree::ptree empty_pt;
```

File: src/runtime_src/core/common/info_aie.cpp [hunk 4]
```diff
@@ -531,6 +531,52 @@ populate_aie_core_gmio(const boost::property_tree::ptree& pt, boost::property_tr
...
+    auto dma_row_it = g_node.second.get_child("dma_rows").begin();
+    for (const auto& node : g_node.second.get_child("dma_columns", empty_pt)) {
+      boost::property_tree::ptree tile;
+      tile.put("column", node.second.data());
+      tile.put("row", dma_row_it->second.data());
+      int row = tile.get<int>("row");
+      int col = tile.get<int>("column");
+
+      // Check whether this core is already added
+      if (is_duplicate_core(tile_array, row, col))
```
Comment (rozumx) [inline src/runtime_src/core/common/info_aie.cpp:570]: It would be better if `tile `was passed in instead of `row `and `col`.

File: src/runtime_src/core/common/info_aie.cpp [hunk 5]
```diff
@@ -531,6 +531,52 @@ populate_aie_core_gmio(const boost::property_tree::ptree& pt, boost::property_tr
...
+        continue;
+
+      populate_aie_core(core_info, tile, row, col);
```
Comment (rozumx) [inline src/runtime_src/core/common/info_aie.cpp:573]: While not part of this PR, you might wish to reconsider refactoring this function for redundant data is being used (e.g., `tile `contains the `row `and `col `values).

In fact, it looks like the logic to determine if the `tile `is already part of `core_info `should be part of `populate_aie_core()`.

File: src/runtime_src/core/tools/common/ReportAie.cpp [hunk 6]
```diff
@@ -133,10 +138,12 @@ ReportAie::writeReport(const xrt_core::device* /*_pDevice*/,
 	if(tile.second.find("dma") != tile.second.not_found()) {
           _output << boost::format("    %s:\n") % "DMA";
 
-          _output << boost::format("%12s:\n") % "FIFO";
-          for(auto& node : tile.second.get_child("dma.fifo.counters")) {
-            _output << fmt16("%s") % node.second.get<std::string>("index")
+	  if(tile.second.find("dma.fifo") != tile.second.not_found()) {
+            _output << boost::format("%12s:\n") % "FIFO";
+            for(auto& node : tile.second.get_child("dma.fifo.counters")) {
```
Comment (rozumx) [inline src/runtime_src/core/tools/common/ReportAie.cpp:143]:
```suggestion
            for(const auto& node : tile.second.get_child("dma.fifo.counters")) {

```

File: src/runtime_src/core/tools/common/ReportAie.cpp [hunk 7]
```diff
@@ -166,7 +173,7 @@ ReportAie::writeReport(const xrt_core::device* /*_pDevice*/,
 
         if(tile.second.find("locks") != tile.second.not_found()) {
           _output << boost::format("    %s:\n") % "Locks";
-          for(auto& node : tile.second.get_child("locks",empty_ptree)) {
+          for(auto& node : tile.second.get_child("locks", empty_ptree)) {
```
Comment (rozumx) [inline src/runtime_src/core/tools/common/ReportAie.cpp:176]:
```suggestion
          for(const auto& node : tile.second.get_child("locks", empty_ptree)) {
```

File: src/runtime_src/core/tools/common/ReportAie.cpp [hunk 8]
```diff
@@ -175,9 +182,9 @@ ReportAie::writeReport(const xrt_core::device* /*_pDevice*/,
 
         if(tile.second.find("errors") != tile.second.not_found()) {
           _output << boost::format("    %s:\n") % "Errors";
-          for(auto& node : tile.second.get_child("errors",empty_ptree)) {
+          for(auto& node : tile.second.get_child("errors", empty_ptree)) {
```
Comment (rozumx) [inline src/runtime_src/core/tools/common/ReportAie.cpp:185]:
```suggestion
          for(const auto& node : tile.second.get_child("errors", empty_ptree)) {

```

File: src/runtime_src/core/tools/common/ReportAie.cpp [hunk 9]
```diff
@@ -175,9 +182,9 @@ ReportAie::writeReport(const xrt_core::device* /*_pDevice*/,
 
         if(tile.second.find("errors") != tile.second.not_found()) {
           _output << boost::format("    %s:\n") % "Errors";
-          for(auto& node : tile.second.get_child("errors",empty_ptree)) {
+          for(auto& node : tile.second.get_child("errors", empty_ptree)) {
             _output << boost::format("        %s:\n") % node.second.get<std::string>("module");
-            for(auto& enode : node.second.get_child("error",empty_ptree)) {
+            for(auto& enode : node.second.get_child("error", empty_ptree)) {
```
Comment (rozumx) [inline src/runtime_src/core/tools/common/ReportAie.cpp:187]:
```suggestion
            for(const auto& enode : node.second.get_child("error", empty_ptree)) {

```

File: src/runtime_src/core/tools/common/ReportAie.cpp [hunk 10]
```diff
@@ -187,21 +194,25 @@ ReportAie::writeReport(const xrt_core::device* /*_pDevice*/,
 
         if(tile.second.find("events") != tile.second.not_found()) {
           _output << boost::format("    %s:\n") % "Events";
-          for(auto& node : tile.second.get_child("events",empty_ptree)) {
+          for(auto& node : tile.second.get_child("events", empty_ptree)) {
```
Comment (rozumx) [inline src/runtime_src/core/tools/common/ReportAie.cpp:197]:
```suggestion
          for(const auto& node : tile.second.get_child("events", empty_ptree)) {

```

File: src/runtime_src/core/tools/common/ReportAie.cpp [hunk 11]
```diff
@@ -187,21 +194,25 @@ ReportAie::writeReport(const xrt_core::device* /*_pDevice*/,
...
         }
       }
-      _output << boost::format("    %s\n") % "Pl Kernel Instances in Graph:";
-      for (auto& node : graph.get_child("pl_kernel")) {
-        _output << boost::format("      %s\n") % node.second.data();
+
+      const boost::property_tree::ptree& pl_kernel = graph.get_child("pl_kernel");
+      if (!pl_kernel.empty()) {
+        _output << boost::format("    %s\n") % "Pl Kernel Instances in Graph:";
+        for (auto& node : graph.get_child("pl_kernel")) {
```
Comment (rozumx) [inline src/runtime_src/core/tools/common/ReportAie.cpp:208]: Curly brace is not needed.

**Review by rozumx — CHANGES_REQUESTED** [review]

File: src/runtime_src/core/common/info_aie.cpp (hunk 2 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/common/info_aie.cpp]: Fixed. Thanks

File: src/runtime_src/core/common/info_aie.cpp (hunk 3 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/common/info_aie.cpp]: Fixed. Thanks

File: src/runtime_src/core/common/info_aie.cpp (hunk 4 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/common/info_aie.cpp]: Fixed. Thanks

File: src/runtime_src/core/common/info_aie.cpp (hunk 5 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/common/info_aie.cpp]: Fixed. Thanks

File: src/runtime_src/core/tools/common/ReportAie.cpp (hunk 7 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/tools/common/ReportAie.cpp]: Fixed. Thanks

File: src/runtime_src/core/tools/common/ReportAie.cpp (hunk 8 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/tools/common/ReportAie.cpp]: Fixed. Thanks

File: src/runtime_src/core/tools/common/ReportAie.cpp (hunk 9 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/tools/common/ReportAie.cpp]: Fixed. Thanks

File: src/runtime_src/core/tools/common/ReportAie.cpp (hunk 10 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/tools/common/ReportAie.cpp]: Fixed. Thanks

File: src/runtime_src/core/tools/common/ReportAie.cpp (hunk 11 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/tools/common/ReportAie.cpp]: Fixed. Thanks

File: src/runtime_src/core/tools/common/ReportAie.cpp (hunk 6 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/tools/common/ReportAie.cpp]: Fixed. Thanks

**Review by chvamshi-xilinx — APPROVED** [review]

**Comment by dayeh-xilinx** [comment]
retest this please.

File: src/runtime_src/core/common/info_aie.cpp [hunk 12]
```diff
@@ -420,14 +420,16 @@ populate_aie_shim(const xrt_core::device *device, const std::string& desc)
         ....
 */
 
-// This funtion populate a specific AIE core given as an input of [row:col]
+// This function populate a specific AIE core given as an input of [row:col]
```
Comment (stsoe) [inline src/runtime_src/core/common/info_aie.cpp:423]: Grammatically incorrect.   Simplify maybe:
```suggestion
// Populate a specific AIE core given as an input of [row:col]
```

File: src/runtime_src/core/common/info_aie.cpp [hunk 13]
```diff
@@ -420,14 +420,16 @@ populate_aie_shim(const xrt_core::device *device, const std::string& desc)
...
-populate_aie_core(const boost::property_tree::ptree& pt_core, boost::property_tree::ptree& tile,
-		  int row, int col)
+populate_aie_core(const boost::property_tree::ptree& pt_core, boost::property_tree::ptree& tile)
 {
   try {
     boost::property_tree::ptree pt;
     boost::property_tree::ptree empty_pt;
+
+    int row = tile.get<int>("row");
+    int col = tile.get<int>("column");
```
Comment (stsoe) [inline src/runtime_src/core/common/info_aie.cpp:432]: More typing yes, but don't repeat the type, it's prone to error.
```suggestion
    auto row = tile.get<int>("row");
    auto col = tile.get<int>("column");
```

File: src/runtime_src/core/common/info_aie.cpp [hunk 14]
```diff
@@ -531,6 +533,49 @@ populate_aie_core_gmio(const boost::property_tree::ptree& pt, boost::property_tr
   pt_array.add_child("gmios",gmio_array);
 }
 
+// This function check the duplicate entry in tile_aaray for the same core [row:col]
+bool
+is_duplicate_core(const boost::property_tree::ptree& tile_array, boost::property_tree::ptree& tile)
+{
+  int row = tile.get<int>("row");
+  int col = tile.get<int>("column");
```
Comment (stsoe) [inline src/runtime_src/core/common/info_aie.cpp:541]:
```suggestion
  auto row = tile.get<int>("row");
  auto col = tile.get<int>("column");
```

**Review by stsoe — CHANGES_REQUESTED** [review]

File: src/runtime_src/core/common/info_aie.cpp (hunk 12 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/common/info_aie.cpp]: Fixed. Thanks

File: src/runtime_src/core/common/info_aie.cpp [hunk 15]
```diff
@@ -531,6 +533,49 @@ populate_aie_core_gmio(const boost::property_tree::ptree& pt, boost::property_tr
   pt_array.add_child("gmios",gmio_array);
 }
 
+// This function check the duplicate entry in tile_aaray for the same core [row:col]
```
[the remark this answers is not in the document]
Comment (saifuddin-xilinx) [reply src/runtime_src/core/common/info_aie.cpp:536]: Fixed. Thanks

**Review by stsoe — APPROVED** [review]

**Review by AShivangi — APPROVED** [review]

**Review by stsoe — APPROVED** [review]

**Review by AShivangi — APPROVED** [review]

File: src/runtime_src/core/common/info_aie.cpp [hunk 16]
```diff
@@ -531,6 +533,49 @@ populate_aie_core_gmio(const boost::property_tree::ptree& pt, boost::property_tr
   pt_array.add_child("gmios",gmio_array);
 }
 
+// Check for duplicate entry in tile_array for the same core [row:col]
+bool
+is_duplicate_core(const boost::property_tree::ptree& tile_array, boost::property_tree::ptree& tile)
+{
+  auto row = tile.get<int>("row");
```
Comment (rozumx) [inline src/runtime_src/core/common/info_aie.cpp:540]:
```suggestion
  const auto row = tile.get<int>("row");
```

File: src/runtime_src/core/common/info_aie.cpp [hunk 17]
```diff
@@ -531,6 +533,49 @@ populate_aie_core_gmio(const boost::property_tree::ptree& pt, boost::property_tr
...
+// Populate a specific AIE core which is unused but memory buffers exist [row:col]
+void
+populate_buffer_only_cores(const boost::property_tree::ptree& pt,
+			  const boost::property_tree::ptree& core_info, int gr_id,
+			  boost::property_tree::ptree& tile_array)
+{
+  const boost::property_tree::ptree empty_pt;
+
+  for (const auto& g_node : pt.get_child("aie_metadata.EventGraphs", empty_pt)) {
+    if (gr_id != g_node.second.get<int>("id"))
```
[the remark this answers is not in the document]
Comment (saifuddin-xilinx) [reply src/runtime_src/core/common/info_aie.cpp:559]: Thanks.

File: src/runtime_src/core/common/info_aie.cpp [hunk 18]
```diff
@@ -531,6 +533,49 @@ populate_aie_core_gmio(const boost::property_tree::ptree& pt, boost::property_tr
...
 }
 
+// Check for duplicate entry in tile_array for the same core [row:col]
+bool
+is_duplicate_core(const boost::property_tree::ptree& tile_array, boost::property_tree::ptree& tile)
+{
+  auto row = tile.get<int>("row");
+  auto col = tile.get<int>("column");
+  for (auto& node : tile_array) {
+      if ((node.second.get<int>("column") == col) && (node.second.get<int>("row") == row))
```
[the remark this answers is not in the document]
Comment (saifuddin-xilinx) [reply src/runtime_src/core/common/info_aie.cpp]: Fixed

File: src/runtime_src/core/tools/common/ReportAie.cpp [hunk 19]
```diff
@@ -187,21 +194,24 @@ ReportAie::writeReport(const xrt_core::device* /*_pDevice*/,
...
           _output << std::endl;
         }
       }
-      _output << boost::format("    %s\n") % "Pl Kernel Instances in Graph:";
-      for (auto& node : graph.get_child("pl_kernel")) {
-        _output << boost::format("      %s\n") % node.second.data();
+
+      const boost::property_tree::ptree& pl_kernel = graph.get_child("pl_kernel");
+      if (!pl_kernel.empty()) {
+        _output << boost::format("    %s\n") % "Pl Kernel Instances in Graph:";
```
[the remark this answers is not in the document]
Comment (saifuddin-xilinx) [reply src/runtime_src/core/tools/common/ReportAie.cpp:207]: I believe, the fix is required for all over ReportAie.cpp and ReportAieShim.cpp. Which is a big change and also need to verify all the outputs. 
I have created a separate CR for 2022.2 to fix this formatting issue. 
https://jira.xilinx.com/browse/CR-1125796

File: src/runtime_src/core/common/info_aie.cpp [hunk 20]
```diff
@@ -531,6 +533,49 @@ populate_aie_core_gmio(const boost::property_tree::ptree& pt, boost::property_tr
...
+    for (const auto& node : g_node.second.get_child("dma_columns", empty_pt)) {
+      boost::property_tree::ptree tile;
+      tile.put("column", node.second.data());
+      tile.put("row", dma_row_it->second.data());
+      // Check whether this core is already added
+      if (is_duplicate_core(tile_array, tile))
+        continue;
+
+      populate_aie_core(core_info, tile);
+      dma_row_it++;
```
[the remark this answers is not in the document]
Comment (saifuddin-xilinx) [reply src/runtime_src/core/common/info_aie.cpp]: Added the check. Thanks

File: src/runtime_src/core/tools/common/ReportAie.cpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (rozumx) [reply src/runtime_src/core/tools/common/ReportAie.cpp:207]: Thank you for creating the CR.  And yes there is a lot of changes.  Good news is that I am not asking to do that over the entire code base, just for the code that was modified.  Cut-n-paste bad code only causes more bad code to propagate, hence my request to clean up this "new" code.

That said, being that we are at the _end of the release_ and that you have created a CR, I am O.K. to overlook this issue.  Especially since, make it now could introduce a bug / issue.

**Review by rozumx — APPROVED** [review]
